logic/RuleImpl.java

```diff
@@ -23,164 +23,73 @@
 
 import com.vaticle.typedb.client.api.TypeDBTransaction;
 import com.vaticle.typedb.client.api.logic.Rule;
-import com.vaticle.typedb.client.common.exception.TypeDBClientException;
-import com.vaticle.typedb.protocol.LogicProto;
+import com.vaticle.typedb.client.common.NativeObject;
 import com.vaticle.typeql.lang.TypeQL;
 import com.vaticle.typeql.lang.pattern.Conjunction;
 import com.vaticle.typeql.lang.pattern.Pattern;
 import com.vaticle.typeql.lang.pattern.variable.ThingVariable;
 
-import java.util.Objects;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_delete;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_get_label;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_get_then;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_get_when;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_is_deleted;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_set_label;
+import static com.vaticle.typedb.client.jni.typedb_client.rule_to_string;
 
-import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.MISSING_LABEL;
-import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.MISSING_TRANSACTION;
-import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Rule.deleteReq;
-import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Rule.setLabelReq;
-import static com.vaticle.typedb.common.util.Objects.className;
-
-public class RuleImpl implements Rule {
-
-    private final String label;
-    private final Conjunction<? extends Pattern> when;
-    private final ThingVariable<?> then;
+public class RuleImpl extends NativeObject<com.vaticle.typedb.client.jni.Rule> implements Rule {
     private final int hash;
 
-    RuleImpl(String label, Conjunction<? extends Pattern> when, ThingVariable<?> then) {
-        if (label == null || label.isEmpty()) throw new TypeDBClientException(MISSING_LABEL);
-        this.label = label;
-        this.when = when;
-        this.then = then;
-        this.hash = Objects.hash(this.label);
-    }
-
-    public static RuleImpl of(LogicProto.Rule ruleProto) {
-        return new RuleImpl(
-                ruleProto.getLabel(),
-                TypeQL.parsePattern(ruleProto.getWhen()).asConjunction(),
-                TypeQL.parseVariable(ruleProto.getThen()).asThing()
-        );
+    RuleImpl(com.vaticle.typedb.client.jni.Rule rule) {
+        super(rule);
+        this.hash = toString().hashCode();
     }
 
     @Override
     public String getLabel() {
-        return label;
+        return rule_get_label(nativeObject);
     }
 
     @Override
     public Conjunction<? extends Pattern> getWhen() {
-        return when;
+        return TypeQL.parsePattern(rule_get_when(nativeObject)).asConjunction();
```

Comment:
I wonder if we  should establish a pattern of caching these moderately expensive lookup+computation steps. I think users would think of these as simple getters that should be 'free', so I think we should...?